depinject/container.go

```diff
@@ -130,7 +132,57 @@ func (c *container) getResolver(typ reflect.Type, key string) (resolver, error)
 		c.resolvers[mapType] = &mapOfOnePerModuleResolver{r}
 	}
 
-	return c.resolvers[typ], nil
+	res := c.resolvers[typ]
+
+	if res == nil && typ.Kind() == reflect.Interface {
+		var matches []reflect.Type
+		for k := range c.resolvers {
+			if k.Kind() != reflect.Interface && k.Implements(typ) {
+				matches = append(matches, k)
+			}
+		}
+
+		if len(matches) == 1 {
+			res = c.resolvers[matches[0]]
+			c.logf("Implicitly registering resolver %v for interface type %v", matches[0], typ)
+			c.resolvers[typ] = res
+		} else if len(matches) > 1 {
+			return nil, &ErrMultipleImplicitInterfaceBindings{Interface: typ, Matches: matches}
+		}
+	}
+
+	return res, nil
+}
+
+func (c *container) getPreferredResolver(typ reflect.Type, key *moduleKey) (resolver, error) {
+	var pref preference
+	var found bool
+
+	// module scoped binding takes precedence
+	pref, found = c.preferences[preferenceKeyFromType(typ, key)]
+
+	// fallback to global scope binding
+	if !found {
+		pref, found = c.preferences[preferenceKeyFromType(typ, nil)]
+	}
+
+	if !found {
+		return nil, nil
+	}
+
+	if pref.resolver != nil {
+		return pref.resolver, nil
+	}
+
+	for k, res := range c.resolvers {
```

Comment:
Looks like resolvers should just be string keyed so we don't have to do this loop search

Reply:
linear search is only required for explicit binding. otherwise we just use a map